src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,885 @@
+                                    permittedJumps |= PermittedJumps.Break | PermittedJumps.Continue;
+                                }
+                            }
+                            break;
+                    }
+                }
+
+                switch (n.kind) {
+                    case SyntaxKind.ThisType:
+                    case SyntaxKind.ThisKeyword:
+                        facts |= RangeFacts.UsesThis;
+                        break;
+                    case SyntaxKind.LabeledStatement:
+                        {
+                            const label = (<LabeledStatement>n).label;
+                            (seenLabels || (seenLabels = [])).push(label.text);
+                            forEachChild(n, visit);
+                            seenLabels.pop();
+                            break;
+                        }
+                    case SyntaxKind.BreakStatement:
+                    case SyntaxKind.ContinueStatement:
+                        {
+                            const label = (<BreakStatement | ContinueStatement>n).label;
+                            if (label) {
+                                if (!contains(seenLabels, label.text)) {
+                                    // attempts to jump to label that is not in range to be extracted
+                                    (errors || (errors = [])).push(createDiagnosticForNode(n, Messages.CannotExtractRangeContainingLabeledBreakOrContinueStatementWithTargetOutsideOfTheRange));
+                                }
+                            }
+                            else {
+                                if (!(permittedJumps & (SyntaxKind.BreakStatement ? PermittedJumps.Break : PermittedJumps.Continue))) {
+                                    // attempt to break or continue in a forbidden context
+                                    (errors || (errors = [])).push(createDiagnosticForNode(n, Messages.CannotExtractRangeContainingConditionalBreakOrContinueStatements));
+                                }
+                            }
+                            break;
+                        }
+                    case SyntaxKind.AwaitExpression:
+                        facts |= RangeFacts.IsAsyncFunction;
+                        break;
+                    case SyntaxKind.YieldExpression:
+                        facts |= RangeFacts.IsGenerator;
+                        break;
+                    case SyntaxKind.ReturnStatement:
+                        if (permittedJumps & PermittedJumps.Return) {
+                            facts |= RangeFacts.HasReturn;
+                        }
+                        else {
+                            (errors || (errors = [])).push(createDiagnosticForNode(n, Messages.CannotExtractRangeContainingConditionalReturnStatement));
+                        }
+                        break;
+                    default:
+                        forEachChild(n, visit);

```

Comment:
What are we using the diagnostics for?  Previewing them to the user?  The comment above suggests they are not "user-facing".